src/services/functionAppService.test.ts

```diff
@@ -0,0 +1,165 @@
+import mockFs from "mock-fs"
+import { MockFactory } from "../test/mockFactory";
+import Serverless from "serverless";
+import { FunctionAppService } from "./functionAppService";
+import { constants } from "../config"
+
+jest.mock("@azure/arm-resources")
+
+jest.mock("@azure/arm-appservice")
+import { WebSiteManagementClient } from "@azure/arm-appservice";
+
+import axios from "axios";
+import MockAdapter from "axios-mock-adapter";
+
+describe("Function App Service", () => {
+  
+  const app = MockFactory.createTestFunctionApp();
+  const slsService = MockFactory.createTestService();
+  const variables = MockFactory.createTestVariables();
+  const provider = MockFactory.createTestAzureServiceProvider();
+
+  let sendFile;
+  let webAppDelete;
+
+  const masterKey = "masterKey";
+  const authKey = "authKey";
+  const syncTriggersMessage = "sync triggers success";
+  const functions = MockFactory.createTestFunctions();
+
+  const masterKeyUrl = `https://${app.defaultHostName}/admin/host/systemkeys/_master`;
+  const authKeyUrl = `https://management.azure.com${app.id}/functions/admin/token?api-version=2016-08-01`;
+  const syncTriggersUrl = `https://management.azure.com${app.id}/syncfunctiontriggers?api-version=2016-08-01`;
+  const listFunctionsUrl = `https://management.azure.com${app.id}/functions?api-version=2016-08-01`;
+  const uploadUrl = `https://${app.name}${constants.scmDomain}${constants.scmZipDeployApiPath}`
+
+  beforeAll(() => {   
+
+    // TODO: How to spy on defaul exported function?
+    const axiosMock = new MockAdapter(axios);
+    
+    // Master Key
+    axiosMock.onGet(masterKeyUrl).reply(200, { value: masterKey });
+    // Auth Key
+    axiosMock.onGet(authKeyUrl).reply(200, authKey);  
+    // Sync Triggers
+    axiosMock.onPost(syncTriggersUrl).reply(200, syncTriggersMessage);
+    // List Functions
+    axiosMock.onGet(listFunctionsUrl).reply(200, { value: functions });
+
+    mockFs({
+      "app.zip": "contents",
+    }, {createCwd: true, createTmp: true});
+  });
+
+  beforeEach(() => {
+    webAppDelete = jest.fn();
+    sendFile = jest.fn();
+
+    WebSiteManagementClient.prototype.webApps = {
```

Comment:
These methods can be mocked by mocking the sub object directly.  ex) `WebApp.prototype.get = jest.fn(...)`